backend/dataline/services/llm_flow/nodes.py

```diff
@@ -60,106 +56,53 @@ def run(cls, state: QueryGraphState) -> QueryGraphStateUpdate:
-                messages.append(function_message)
             else:
-                # Get data description from results
-                if response.columns and response.rows:
-                    data_types = [type(cell).__name__ for cell in response.rows[0]]
-                    data_description = (
-                        f"Returned data description:\nColumns:{response.columns}\nFirst row: {data_types}"
-                    )
-                elif len(response.rows) == 1:
-                    data_types = [type(cell).__name__ for cell in response.rows[0]]
-                    data_description = f"Returned data description:\nOnly one row: {data_types}"
-                else:
-                    data_description = "No data returned"
-
-                function_message = FunctionMessage(
-                    content="Query executed successfully - here is the returned data description. "
-                    "I cannot view the data for security reasons but the user should be able to see the results!\n"
-                    f"{data_description}",
-                    name=action.tool,
+                # We call the tool_executor and get back a response
+                response = state.tool_executor.tool_map[tool_call["name"]].run(tool_call["args"])
+                # We use the response to create a ToolMessage
+                tool_message = ToolMessage(
+                    content=str(response), name=tool_call["name"], tool_call_id=str(tool_call["id"])
                 )
-                messages.append(function_message)
-        else:
-            # We call the tool_executor and get back a response
-            response = state.tool_executor.invoke(action)
-            # We use the response to create a FunctionMessage
-            function_message = FunctionMessage(content=str(response), name=action.tool)
-            messages.append(function_message)
+                output_messages.append(tool_message)
 
         # We return a list, because this will get added to the existing list
-        return state_update(messages=messages, results=results)
+        return state_update(messages=output_messages, results=results)
 
 
 class CallListTablesToolNode(Node):
     __name__ = "get_tables"
 
     @classmethod
     def run(cls, state: QueryGraphState) -> QueryGraphStateUpdate:
-        action = ToolInvocation(tool=SQLToolNames.LIST_SQL_TABLES, tool_input={})
-        response = cast(SelectedTablesResult, state.tool_executor.invoke(action))
-        function_message = FunctionMessage(content=str(", ".join(response.tables)), name=action.tool)
-        return state_update(messages=[function_message], results=[response])
+        # action = ToolInvocation(tool=SQLToolNames.LIST_SQL_TABLES, tool_input={})
+        tool = cast(ListSQLTablesTool, state.tool_executor.tool_map[SQLToolNames.LIST_SQL_TABLES])
+        response: list[str] = tool.run(tool_input={})
+        tool_message = FunctionMessage(content=str(", ".join(response)), name=tool.name)
```

Comment:
I don't really get the difference between FunctionMessage and ToolMessage?

Reply:
Is it that one is called by the LLM (Tool, hence tool_id) and the other is called by our code (Function)?